pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,777 @@
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curIPs); err != nil {
+			return err
+		}
+		if err = c.deleteInterface(preIfName); err != nil {
+			return err
+		}
+	} else if (!reflect.DeepEqual(preIPs, curIPs)) || (preEEName != curEEName) {
+		klog.InfoS("Found interface configuration is changed", "preIPs", preIPs, "preEEName", preEEName,
+			"curIPs", curIPs, "curEEName", curEEName)
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curIPs); err != nil {
+			return err
+		}
+	}
+	c.syncedExternalNodes.Add(curEn)
+	// Notify the ExternalNode update event to NetworkPolicyController.
+	c.nodeUpdateNotifier.Notify(v1beta2.ExternalEntityReference{
+		Name:      curEEName,
+		Namespace: curEn.Namespace,
+	})
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalNode(key string) error {
+	klog.InfoS("Deleting ExternalNode", "key", key)
+	obj, exists, _ := c.syncedExternalNodes.GetByKey(key)
+	if !exists {
+		klog.InfoS("Skipping ExternalNode deletion as it hasn't been synced", "ExternalEntityKey", key)
+		return nil
+	}
+	en := obj.(*v1alpha1.ExternalNode)
+	ifName, _, err := getHostInterfaceName(en.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	if err := c.deleteInterface(ifName); err != nil {
+		return err
+	}
+	c.syncedExternalNodes.Delete(en)
+	return nil
+}
+
+func (c *ExternalNodeController) deleteInterface(ifName string) error {
+	hostIface, ifaceExists := c.ifaceStore.GetInterfaceByName(ifName)
+	if !ifaceExists {
+		klog.InfoS("Skipping deleting host interface since it doesn't exist ", "ifName", ifName)
+		return nil
+	}
+	klog.InfoS("Deleting interface", "ifName", ifName)
+	if err := c.removeOVSPortsAndFlows(hostIface); err != nil {
+		return err
+	}
+	c.ifaceStore.DeleteInterface(hostIface)
+	return nil
+}
+
+func (c *ExternalNodeController) createOVSPortsAndFlows(uplinkName, hostIfName, eeNamespace string, eeName string, ips []string) (*interfacestore.InterfaceConfig, error) {
+	adapterConfig, err := util.GetUplinkConfig(hostIfName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get the configuration on the host interface", "hostInterface", hostIfName)
+		return nil, err
+	}
+	if err := util.RenameInterface(hostIfName, uplinkName); err != nil {
+		klog.ErrorS(err, "Failed to rename host interface name to uplink name", "hostInterface", hostIfName, "uplink", uplinkName)
+		return nil, err
+	}
+	success := false
+	defer func() {
+		if !success {
+			util.RenameInterface(uplinkName, hostIfName)
+		}
+	}()
+	// Create uplink port on OVS.
+	uplinkExternalIDs := map[string]interface{}{
+		interfacestore.AntreaInterfaceTypeKey: interfacestore.AntreaUplink,
+	}
+	uplinkUUID, err := c.ovsBridgeClient.CreatePort(uplinkName, uplinkName, uplinkExternalIDs)
+	if err != nil {
+		klog.ErrorS(err, "Failed to create uplink port on OVS", "uplink", uplinkName)
+		return nil, err
+	}
+	defer func() {
+		if !success {
+			c.ovsBridgeClient.DeletePort(uplinkUUID)
```

Comment:
Log the error if any of rollback funcs fails.

And what will happen if rollback fails? Will the retry or deletion be able to resume the creation or clean up the state?

Reply:
I have added the logs here as suggested. Since it will return error to its caller which finally is function syncExternalNode, I think it should be put back to the work queue and retry.